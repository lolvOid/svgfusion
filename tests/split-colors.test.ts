import {
  extractColors,
  replaceColorsWithProps,
} from '../src/utils/color-extractor';

describe('Split Colors Feature', () => {
  // Test 1: Basic color extraction
  test('should extract fill and stroke colors separately', () => {
    const svg = `<svg><path fill="#FF0000" stroke="#00FF00" /></svg>`;
    const colorInfo = extractColors(svg);

    expect(colorInfo.fillColors).toEqual(['#FF0000']);
    expect(colorInfo.strokeColors).toEqual(['#00FF00']);
    expect(colorInfo.fillColorMap.get('#FF0000')).toBe('fillColor1');
    expect(colorInfo.strokeColorMap.get('#00FF00')).toBe('strokeColor1');

    // Snapshot test for color extraction result
    expect(colorInfo).toMatchSnapshot();
  });

  // Test 2: Gradient color extraction
  test('should extract gradient stop colors', () => {
    const svg = `<svg><linearGradient><stop stop-color="#FF0000"/><stop stop-color="#0000FF"/></linearGradient></svg>`;
    const colorInfo = extractColors(svg);

    expect(colorInfo.gradientColors).toEqual(['#0000FF', '#FF0000']);
    expect(colorInfo.gradientColorMap.get('#FF0000')).toBe('gradientColor2');
    expect(colorInfo.gradientColorMap.get('#0000FF')).toBe('gradientColor1');

    // Snapshot test for gradient color extraction
    expect(colorInfo).toMatchSnapshot();
  });

  // Test 3: Same color in different contexts
  test('should handle same color in fill and stroke separately', () => {
    const svg = `<svg><path fill="#000000" stroke="#000000" /></svg>`;
    const colorInfo = extractColors(svg);

    expect(colorInfo.fillColors).toEqual(['#000000']);
    expect(colorInfo.strokeColors).toEqual(['#000000']);
    expect(colorInfo.fillColorMap.get('#000000')).toBe('fillColor1');
    expect(colorInfo.strokeColorMap.get('#000000')).toBe('strokeColor1');

    // Snapshot test for same color in different contexts
    expect(colorInfo).toMatchSnapshot();
  });

  // Test 4: React prop replacement
  test('should replace colors with React props', () => {
    const svg = `<path fill="#FF0000" stroke="#00FF00" />`;
    const colorInfo = extractColors(svg);
    const result = replaceColorsWithProps(svg, colorInfo, 'react');

    expect(result).toContain('fill={fillColor1}');
    expect(result).toContain('stroke={strokeColor1}');
    expect(result).toContain('className={fillColor1Class}');
    expect(result).toContain('className={strokeColor1Class}');

    // Snapshot test for React prop replacement
    expect(result).toMatchSnapshot();
  });

  // Test 5: Vue prop replacement
  test('should replace colors with Vue props', () => {
    const svg = `<stop stop-color="#FF0000" />`;
    const colorInfo = extractColors(svg);
    const result = replaceColorsWithProps(svg, colorInfo, 'vue');

    expect(result).toContain(':stop-color="gradientColor1"');
    expect(result).toContain(':class="gradientColor1Class"');

    // Snapshot test for Vue prop replacement
    expect(result).toMatchSnapshot();
  });

  // Test 6: Complex SVG with mixed colors
  test('should handle complex SVG with mixed fill, stroke, and gradient colors', () => {
    const svg = `
      <svg>
        <path fill="#FF0000" stroke="#00FF00" />
        <circle fill="#0000FF" stroke="#FF0000" />
        <linearGradient>
          <stop stop-color="#FFFF00" />
          <stop stop-color="#FF00FF" />
        </linearGradient>
      </svg>
    `;
    const colorInfo = extractColors(svg);

    // Snapshot test for complex color extraction
    expect(colorInfo).toMatchSnapshot();
  });

  // Test 7: Edge cases with invalid colors
  test('should handle invalid and special color values', () => {
    const svg = `
      <svg>
        <path fill="none" stroke="transparent" />
        <circle fill="currentColor" stroke="#invalid" />
        <rect fill="rgb(255, 0, 0)" stroke="hsl(120, 100%, 50%)" />
      </svg>
    `;
    const colorInfo = extractColors(svg);

    // Snapshot test for edge cases
    expect(colorInfo).toMatchSnapshot();
  });
});
