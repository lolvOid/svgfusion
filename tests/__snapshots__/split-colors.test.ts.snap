// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Split Colors Feature should extract fill and stroke colors separately 1`] = `
{
  "colorMap": Map {
    "#00FF00" => "color1",
    "#FF0000" => "color2",
  },
  "colors": [
    "#00FF00",
    "#FF0000",
  ],
  "fillColorMap": Map {
    "#FF0000" => "fillColor1",
  },
  "fillColors": [
    "#FF0000",
  ],
  "gradientColorMap": Map {},
  "gradientColors": [],
  "strokeColorMap": Map {
    "#00FF00" => "strokeColor1",
  },
  "strokeColors": [
    "#00FF00",
  ],
}
`;

exports[`Split Colors Feature should extract gradient stop colors 1`] = `
{
  "colorMap": Map {
    "#0000FF" => "color1",
    "#FF0000" => "color2",
  },
  "colors": [
    "#0000FF",
    "#FF0000",
  ],
  "fillColorMap": Map {},
  "fillColors": [],
  "gradientColorMap": Map {
    "#0000FF" => "gradientColor1",
    "#FF0000" => "gradientColor2",
  },
  "gradientColors": [
    "#0000FF",
    "#FF0000",
  ],
  "strokeColorMap": Map {},
  "strokeColors": [],
}
`;

exports[`Split Colors Feature should handle complex SVG with mixed fill, stroke, and gradient colors 1`] = `
{
  "colorMap": Map {
    "#0000FF" => "color1",
    "#00FF00" => "color2",
    "#FF0000" => "color3",
    "#FF00FF" => "color4",
    "#FFFF00" => "color5",
  },
  "colors": [
    "#0000FF",
    "#00FF00",
    "#FF0000",
    "#FF00FF",
    "#FFFF00",
  ],
  "fillColorMap": Map {
    "#0000FF" => "fillColor1",
    "#FF0000" => "fillColor2",
  },
  "fillColors": [
    "#0000FF",
    "#FF0000",
  ],
  "gradientColorMap": Map {
    "#FF00FF" => "gradientColor1",
    "#FFFF00" => "gradientColor2",
  },
  "gradientColors": [
    "#FF00FF",
    "#FFFF00",
  ],
  "strokeColorMap": Map {
    "#00FF00" => "strokeColor1",
    "#FF0000" => "strokeColor2",
  },
  "strokeColors": [
    "#00FF00",
    "#FF0000",
  ],
}
`;

exports[`Split Colors Feature should handle invalid and special color values 1`] = `
{
  "colorMap": Map {
    "hsl(120, 100%, 50%)" => "color1",
    "rgb(255, 0, 0)" => "color2",
  },
  "colors": [
    "hsl(120, 100%, 50%)",
    "rgb(255, 0, 0)",
  ],
  "fillColorMap": Map {
    "rgb(255, 0, 0)" => "fillColor1",
  },
  "fillColors": [
    "rgb(255, 0, 0)",
  ],
  "gradientColorMap": Map {},
  "gradientColors": [],
  "strokeColorMap": Map {
    "hsl(120, 100%, 50%)" => "strokeColor1",
  },
  "strokeColors": [
    "hsl(120, 100%, 50%)",
  ],
}
`;

exports[`Split Colors Feature should handle same color in fill and stroke separately 1`] = `
{
  "colorMap": Map {
    "#000000" => "color1",
  },
  "colors": [
    "#000000",
  ],
  "fillColorMap": Map {
    "#000000" => "fillColor1",
  },
  "fillColors": [
    "#000000",
  ],
  "gradientColorMap": Map {},
  "gradientColors": [],
  "strokeColorMap": Map {
    "#000000" => "strokeColor1",
  },
  "strokeColors": [
    "#000000",
  ],
}
`;

exports[`Split Colors Feature should replace colors with React props 1`] = `"<path fill={fillColor1} className={fillColor1Class} stroke={strokeColor1} className={strokeColor1Class} />"`;

exports[`Split Colors Feature should replace colors with Vue props 1`] = `"<stop :stop-color="gradientColor1" :class="gradientColor1Class" />"`;
